auth/services/selfsigned/validator.go

```diff
@@ -19,42 +19,170 @@
+	credentialSubject, proof, err := v.verifyVP(vp, validAt)
+	if err != nil {
+		if errors.As(err, &verificationError{}) {
+			result.InvalidReason = err.Error()
+			return result, nil
+		}
+		return nil, err
+	}
+
+	// 2. Contract validation
+	if proof.Challenge == nil {
+		result.InvalidReason = "challenge is required"
+		return result, nil
+	}
+
+	c, err := contract.ParseContractString(*proof.Challenge, v.validContracts)
+	if err != nil {
+		result.InvalidReason = err.Error()
+		return result, nil
+	}
+	t := time.Now()
+	if validAt != nil {
+		t = *validAt
+	}
+	err = c.VerifyForGivenTime(t)
+	if err != nil {
+		result.InvalidReason = err.Error()
+		return result, nil
+	}
+
+	// 3. check for mandatory attributes in credentialSubject
+	if err = validateRequiredAttributes(credentialSubject); err != nil {
+		result.InvalidReason = err.Error()
+		return result, nil
+	}
+
+	// TODO add role? See #2047
+	disclosedAttributes := map[string]string{
+		services.InitialsTokenClaim:   credentialSubject.Member.Member.Initials,
+		services.FamilyNameTokenClaim: credentialSubject.Member.Member.FamilyName,
+		services.UsernameClaim:        credentialSubject.Member.Identifier,
+		services.EidasIALClaim:        "low",
+	}
+
+	return selfsignedVerificationResult{
+		Status: contract.Valid,
+		// extract organization attributes and add them to the result
+		contractAttributes:  c.Params,
+		disclosedAttributes: disclosedAttributes,
+	}, nil
+}
+
+func (v service) verifyVP(vp vc.VerifiablePresentation, validAt *time.Time) (credentialSubject employeeIdentityCredentialSubject, proof vc.JSONWebSignature2020Proof, resultErr error) {
+	vcs, err := v.vcr.Verifier().VerifyVP(vp, true, validAt)
+	if err != nil {
+		if errors.As(err, &verifier.VerificationError{}) {
+			resultErr = newVerificationError(err.Error())
+			return
+		}
+		resultErr = err
+		return
+	}
+
+	if len(vcs) != 1 {
+		resultErr = newVerificationError("exactly 1 EmployeeIdentityCredential is required")
+		return
+	}
+	if len(vp.Proof) != 1 {
+		resultErr = newVerificationError("exactly 1 Proof is required")
+		return
+	}
+	bytes, _ := json.Marshal(vp.Proof[0])
+	_ = json.Unmarshal(bytes, &proof)
+	if proof.Proof.Type != ssi.JsonWebSignature2020 {
+		resultErr = newVerificationError("proof must be of type JsonWebSignature2020")
+		return
+	}
+	vc := vcs[0]
+	signingMethod := proof.VerificationMethod
+	signingMethod.Fragment = ""
+	if vc.Issuer.String() != signingMethod.String() {
+		resultErr = newVerificationError("signer must be credential issuer")
+		return
+	}
+	bytes, _ = json.Marshal(vc.CredentialSubject[0])
+	_ = json.Unmarshal(bytes, &credentialSubject)
+	if vc.Issuer.String() != credentialSubject.ID {
+		resultErr = newVerificationError("signer must be credentialSubject")
+		return
+	}
+
+	return credentialSubject, proof, nil
+}
+
+func validateRequiredAttributes(credentialSubject employeeIdentityCredentialSubject) error {
+	// check for mandatory attrs
```

Comment:
shouldn't these be validated on the VC already?

Reply:
These are the requirements from the spec (rfc that has to be written). It doesn't validate stuff like issuer, context and id but mostly `credentialSubject` requirements.

Or do you mean it should be under vcr/credential/validator? << where credentials are validated that enter the server wallet from the network?

Reply:
The last one, indeed. I could've been more clear.

Reply:
So we validate identity means (and VPs) inside auth. The attestations (VCs over network) are validated within VCR.